File: frontend/src/phoneNumbers.js
import { parsePhoneNumberFromString } from "libphonenumber-js";

const DECIMAL_DIGIT_SETS = [
  "0123456789",
  "٠١٢٣٤٥٦٧٨٩",
  "۰۱۲۳۴۵۶۷۸۹",
];

export function normalizePhoneDigits(value) {
  return String(value ?? "").normalize("NFKC").replace(/\p{Nd}/gu, (digit) => {
    for (const set of DECIMAL_DIGIT_SETS) {
      const index = set.indexOf(digit);
      if (index >= 0) return String(index);
    }
    return digit;
  });
}

function internationalInput(value) {
  const normalized = normalizePhoneDigits(value).trim();
  return normalized.startsWith("00") ? `+${normalized.slice(2)}` : normalized;
}

export function normalizeInternationalPhone(value) {
  const input = internationalInput(value);
  if (!input.startsWith("+")) {
    throw new Error("Enter the phone number in international format beginning with +.");
  }
  const phone = parsePhoneNumberFromString(input);
  if (!phone?.isValid()) throw new Error("Enter a valid international phone number.");
  return phone.number;
}

export function tryNormalizeInternationalPhone(value) {
  try {
    return normalizeInternationalPhone(value);
  } catch {
    return null;
  }
}

export function normalizeNationalPhone(country, value) {
  if (!country) throw new Error("Choose a valid phone country or region.");
  const input = internationalInput(value);
  const phone = parsePhoneNumberFromString(input, input.startsWith("+") ? undefined : country.region);
  if (!phone?.isValid()) throw new Error("Enter a valid phone number for the selected country or region.");
  if (`+${phone.countryCallingCode}` !== country.code) {
    throw new Error("The phone number country code must match the selected country or region.");
  }
  return {
    country_calling_code: country.code,
    phone_number: phone.nationalNumber,
    phone_e164: phone.number,
  };
}

File: backend/health_hub/phone_numbers.py
import re
import unicodedata

import phonenumbers
from phonenumbers import NumberParseException, PhoneNumberFormat


CALLING_CODE_PATTERN = re.compile(r"^\+[1-9]\d{0,3}$")


def normalize_digits(value):
    normalized = unicodedata.normalize("NFKC", str(value or ""))
    return "".join(
        str(unicodedata.digit(character)) if character.isdigit() else character
        for character in normalized
    )


def _international_input(value):
    normalized = normalize_digits(value).strip()
    if normalized.startswith("00"):
        normalized = f"+{normalized[2:]}"
    return normalized


def _parse_valid_number(value, region=None):
    try:
        number = phonenumbers.parse(value, region)
    except NumberParseException as exc:
        raise ValueError("Enter a valid phone number.") from exc
    if not phonenumbers.is_valid_number(number):
        raise ValueError("Enter a valid phone number.")
    return number


def normalize_international_phone(value):
    raw = _international_input(value)
    if not raw.startswith("+"):
        raise ValueError("Enter the phone number in international format beginning with +.")
    number = _parse_valid_number(raw)
    return phonenumbers.format_number(number, PhoneNumberFormat.E164)


def normalize_patient_phone(country_calling_code, phone_number):
    calling_code = re.sub(r"[\s().-]+", "", normalize_digits(country_calling_code).strip())
    if calling_code.startswith("00"):
        calling_code = f"+{calling_code[2:]}"
    if not CALLING_CODE_PATTERN.fullmatch(calling_code):
        raise ValueError("Choose a valid phone country or region.")

    numeric_calling_code = int(calling_code[1:])
    region = phonenumbers.region_code_for_country_code(numeric_calling_code)
    if region in {"001", "ZZ"}:
        raise ValueError("Choose a valid phone country or region.")

    raw_number = _international_input(phone_number)
    international = raw_number.startswith("+")
    number = _parse_valid_number(raw_number, None if international else region)
    if number.country_code != numeric_calling_code:
        raise ValueError("The phone number country code must match the selected country or region.")

    e164 = phonenumbers.format_number(number, PhoneNumberFormat.E164)
    national_number = e164[len(calling_code):]
    return calling_code, national_number, e164
